rfcs/0084-input-aware-fetchers.md

```diff
@@ -0,0 +1,147 @@
+---
+feature: input-aware-fetchers
+start-date: 2020-12-30
+author: @infinisil
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Change the name of the main fixed-output derivations to contain a hash of all inputs that could affect the output hash, such that if any of them change, a previously-fetched path with the same hash won't be reused. This resolves the long-standing confusion about Nix using an old version of a source even though the url it should be fetched from was updated.
+
+# Motivation
+[motivation]: #motivation
+
+Say we have a `fetchFromGitHub` call like
+```nix
+(import <nixpkgs> {}).fetchFromGitHub {
+  owner = "NixOS";
+  repo = "nix";
+  rev = "2.3.9";
+  sha256 = "0d4c3ddpqa1q4j15cl8d7g3igiw6clqczf8dcp4pbpvlm9a64rki";
+}
+```
+
+With a current nixpkgs version, this can be built successfully, returning the result in store path
+```
+/nix/store/l98gjfznp8lpxi0hvj4i0rw34xnnqma8-source
+```
+
+However, if we now intuitively update the revision of the expression to the new 2.3.10 release like
+```nix
+(import <nixpkgs> {}).fetchFromGitHub {
+  owner = "NixOS";
+  repo = "nix";
+  rev = "2.3.10";
+  sha256 = "0d4c3ddpqa1q4j15cl8d7g3igiw6clqczf8dcp4pbpvlm9a64rki";
+}
+```
+
+We can still build this successfully, getting the exact same store path
+```
+/nix/store/l98gjfznp8lpxi0hvj4i0rw34xnnqma8-source
+```
+
+This can't be right! Of course, people with Nix knowledge will tell you that you need to change the hash to something invalid first in order for it to attempt a refetch. This is because Nix already has a path with that exact output `sha256` hash, so it just reuses that. Indeed, changing the output hash to something invalid works:
```

Comment:
> Of course, people with Nix knowledge will tell you that you need to change the hash to something invalid first in order for it to attempt a refetch.

I'm sorry to hear that this is encouraged as a good practice. Shouldn't people be told instead to use prefetchers whenever updating a source, rather than wastefully downloading the source twice just to get a hash mismatch the first time?

(I concede that I do the invalid-hash procedure myself for Rust packages, for which using a prefetcher would have been beyond my understanding of the Rust packaging functions, but I always use prefetchers for URL sources and Git sources.)

Reply:
It won't download it twice, since it stores the result under the store path for the calculated hash the first time